// @midscene/shared/dist/es/img/info.mjs
import node_assert from "node:assert";
import { Buffer } from "node:buffer";
import get_photon from "./get-photon.mjs";
async function imageInfoOfBase64(imageBase64) {
    const { PhotonImage } = await get_photon();
    const base64Data = imageBase64.replace(/^data:image\/\w+;base64,/, '').replace(/\s/g, '');
    node_assert(base64Data, 'Invalid image: empty base64 data');
    node_assert(/^[A-Za-z0-9+/]+={0,2}$/.test(base64Data) && base64Data.length % 4 !== 1, 'Invalid image: malformed base64 data');
    const imageBuffer = Buffer.from(base64Data, 'base64');
    node_assert(isValidImageBuffer(imageBuffer), 'Invalid image: unsupported format');
    let result;
    try {
        result = PhotonImage.new_from_base64(base64Data);
    } catch (error) {
        throw new Error(`Invalid image: failed to decode base64 data (${error instanceof Error ? error.message : String(error)})`, {
            cause: error
        });
    }
    const image = result instanceof Promise ? await result : result;
    const width = image.get_width();
    const height = image.get_height();
    image.free();
    node_assert(width && height, 'Invalid image: cannot get width or height');
    return {
        width,
        height
    };
}
function isValidPNGImageBuffer(buffer) {
    if (!buffer || buffer.length < 8) return false;
    const isPNG = 0x89 === buffer[0] && 0x50 === buffer[1] && 0x4e === buffer[2] && 0x47 === buffer[3] && 0x0d === buffer[4] && 0x0a === buffer[5] && 0x1a === buffer[6] && 0x0a === buffer[7];
    return isPNG;
}
function isValidJPEGImageBuffer(buffer) {
    if (!buffer || buffer.length < 3) return false;
    return 0xff === buffer[0] && 0xd8 === buffer[1] && 0xff === buffer[2];
}
function isValidImageBuffer(buffer) {
    return isValidPNGImageBuffer(buffer) || isValidJPEGImageBuffer(buffer);
}
function validateScreenshotBuffer(screenshotBuffer, { label, minBufferSize = 0 }) {
    const bufferSize = screenshotBuffer?.length ?? 0;
    if (!screenshotBuffer || 0 === bufferSize) throw new Error(`${label} validation failed: buffer size ${bufferSize} bytes`);
    if (!isValidImageBuffer(screenshotBuffer)) throw new Error(`${label} buffer has invalid image format`);
    if (minBufferSize > 0 && bufferSize < minBufferSize) throw new Error(`${label} validation failed: buffer size ${bufferSize} bytes (minimum: ${minBufferSize})`);
}
export { imageInfoOfBase64, isValidImageBuffer, isValidJPEGImageBuffer, isValidPNGImageBuffer, validateScreenshotBuffer };
