// @yume-chan/scrcpy/esm/2_7/options.js
import { ScrcpyDeviceMessageParsers } from "../base/index.js";
import { AckClipboardHandler, ClipboardStream, ControlMessageTypes, createMediaStreamTransformer, createScrollController, Defaults, parseAudioStreamMetadata, parseDisplay, parseEncoder, parseVideoStreamMetadata, serialize, serializeBackOrScreenOnControlMessage, serializeInjectTouchControlMessage, serializeUHidCreateControlMessage, setListDisplays, setListEncoders, UHidOutputStream, } from "./impl/index.js";
export class ScrcpyOptions2_7 {
    static Defaults = Defaults;
    value;
    get controlMessageTypes() {
        return ControlMessageTypes;
    }
    #clipboard;
    get clipboard() {
        return this.#clipboard;
    }
    #ackClipboardHandler;
    #uHidOutput;
    get uHidOutput() {
        return this.#uHidOutput;
    }
    #deviceMessageParsers = new ScrcpyDeviceMessageParsers();
    get deviceMessageParsers() {
        return this.#deviceMessageParsers;
    }
    constructor(init) {
        this.value = { ...Defaults, ...init };
        if (this.value.videoSource === "camera") {
            this.value.control = false;
        }
        if (this.value.audioDup) {
            this.value.audioSource = "playback";
        }
        if (this.value.control) {
            if (this.value.clipboardAutosync) {
                this.#clipboard = this.#deviceMessageParsers.add(new ClipboardStream());
                this.#ackClipboardHandler = this.#deviceMessageParsers.add(new AckClipboardHandler());
            }
            this.#uHidOutput = this.#deviceMessageParsers.add(new UHidOutputStream());
        }
    }
    serialize() {
        return serialize(this.value, Defaults);
    }
    setListDisplays() {
        setListDisplays(this.value);
    }
    parseDisplay(line) {
        return parseDisplay(line);
    }
    setListEncoders() {
        setListEncoders(this.value);
    }
    parseEncoder(line) {
        return parseEncoder(line);
    }
    parseVideoStreamMetadata(stream) {
        return parseVideoStreamMetadata(this.value, stream);
    }
    parseAudioStreamMetadata(stream) {
        return parseAudioStreamMetadata(stream, this.value);
    }
    createMediaStreamTransformer() {
        return createMediaStreamTransformer(this.value);
    }
    serializeInjectTouchControlMessage(message) {
        return serializeInjectTouchControlMessage(message);
    }
    serializeBackOrScreenOnControlMessage(message) {
        return serializeBackOrScreenOnControlMessage(message);
    }
    serializeSetClipboardControlMessage(message) {
        return this.#ackClipboardHandler.serializeSetClipboardControlMessage(message);
    }
    createScrollController() {
        return createScrollController();
    }
    serializeUHidCreateControlMessage(message) {
        return serializeUHidCreateControlMessage(message);
    }
}
