// @yume-chan/scrcpy/esm/1_24/options.js
import { ScrcpyDeviceMessageParsers } from "../base/index.js";
import { AckClipboardHandler, ClipboardStream, ControlMessageTypes, createMediaStreamTransformer, createScrollController, Defaults, EncoderRegex, parseDisplay, parseEncoder, parseVideoStreamMetadata, serialize, serializeBackOrScreenOnControlMessage, serializeInjectTouchControlMessage, setListDisplays, setListEncoders, } from "./impl/index.js";
export class ScrcpyOptions1_24 {
    static Defaults = Defaults;
    value;
    get controlMessageTypes() {
        return ControlMessageTypes;
    }
    #clipboard;
    get clipboard() {
        return this.#clipboard;
    }
    #ackClipboardHandler;
    #deviceMessageParsers = new ScrcpyDeviceMessageParsers();
    get deviceMessageParsers() {
        return this.#deviceMessageParsers;
    }
    constructor(init) {
        this.value = { ...Defaults, ...init };
        if (this.value.control && this.value.clipboardAutosync) {
            this.#clipboard = this.#deviceMessageParsers.add(new ClipboardStream());
            this.#ackClipboardHandler = this.#deviceMessageParsers.add(new AckClipboardHandler());
        }
    }
    serialize() {
        return serialize(this.value, Defaults);
    }
    setListDisplays() {
        setListDisplays(this.value);
    }
    parseDisplay(line) {
        return parseDisplay(line);
    }
    setListEncoders() {
        setListEncoders(this.value);
    }
    parseEncoder(line) {
        return parseEncoder(line, EncoderRegex);
    }
    parseVideoStreamMetadata(stream) {
        return parseVideoStreamMetadata(this.value, stream);
    }
    createMediaStreamTransformer() {
        return createMediaStreamTransformer(this.value);
    }
    serializeInjectTouchControlMessage(message) {
        return serializeInjectTouchControlMessage(message);
    }
    serializeBackOrScreenOnControlMessage(message) {
        return serializeBackOrScreenOnControlMessage(message);
    }
    serializeSetClipboardControlMessage(message) {
        return this.#ackClipboardHandler.serializeSetClipboardControlMessage(message);
    }
    createScrollController() {
        return createScrollController();
    }
}
