// @yume-chan/scrcpy/esm/1_17/options.js
import { ScrcpyDeviceMessageParsers } from "../base/index.js";
import { ClipboardStream, ControlMessageTypes, createMediaStreamTransformer, createScrollController, Defaults, EncoderRegex, parseDisplay, parseEncoder, parseVideoStreamMetadata, serialize, serializeBackOrScreenOnControlMessage, serializeInjectTouchControlMessage, SerializeOrder, serializeSetClipboardControlMessage, setListDisplays, setListEncoders, } from "./impl/index.js";
export class ScrcpyOptions1_17 {
    static Defaults = Defaults;
    value;
    get controlMessageTypes() {
        return ControlMessageTypes;
    }
    #clipboard;
    get clipboard() {
        return this.#clipboard;
    }
    #deviceMessageParsers = new ScrcpyDeviceMessageParsers();
    get deviceMessageParsers() {
        return this.#deviceMessageParsers;
    }
    constructor(init) {
        this.value = { ...Defaults, ...init };
        if (this.value.control) {
            this.#clipboard = this.#deviceMessageParsers.add(new ClipboardStream());
        }
    }
    serialize() {
        return serialize(this.value, SerializeOrder);
    }
    setListDisplays() {
        setListDisplays(this.value);
    }
    parseDisplay(line) {
        return parseDisplay(line);
    }
    setListEncoders() {
        setListEncoders(this.value);
    }
    parseEncoder(line) {
        return parseEncoder(line, EncoderRegex);
    }
    parseVideoStreamMetadata(stream) {
        return parseVideoStreamMetadata(stream);
    }
    createMediaStreamTransformer() {
        return createMediaStreamTransformer(this.value);
    }
    serializeInjectTouchControlMessage(message) {
        return serializeInjectTouchControlMessage(message);
    }
    serializeBackOrScreenOnControlMessage(message) {
        return serializeBackOrScreenOnControlMessage(message);
    }
    serializeSetClipboardControlMessage(message) {
        return serializeSetClipboardControlMessage(message);
    }
    createScrollController() {
        return createScrollController();
    }
}
